src/scala/com/github/johnynek/bazel_deps/CoursierResolver.scala

```diff
@@ -131,12 +153,125 @@ class CoursierResolver(servers: List[DependencyServer], ec: ExecutionContext, ru
+                      .method("HEAD", HttpRequest.BodyPublishers.noBody)
+                      .build
+                    val resp = httpClient.send(req, HttpResponse.BodyHandlers.discarding)
+                    resp.statusCode() match {
+                      case 404 =>
+                        if (cacheErrors) {
+                          try Files.createFile(errFile(artifactFile).toPath)
+                          catch {
+                            case e: java.nio.file.FileAlreadyExistsException => ()
+                          }
+                        } else {
+                          // println(s"not caching error for $artifact")
+                        }
+                        throw new ArtifactError.NotFound(artifact.url, permanent = Some(true))
+                      case 401 =>
+                        throw new ArtifactError.Unauthorized(artifact.url, realm = None)
+                      case sc if sc >= 400 =>
+                        throw new ArtifactError.DownloadError(s"failed to download: $sc", None)
+                      case _ =>
+                    }
+
+                    val headers = resp.headers
+                    val headersMap = headers.map.asScala.map({ case (k, vs) => (k, vs.asScala) })
+                    val headersJson = headersMap.asJson.noSpaces
+
+                    // Write to a temporary file first.
+                    val writer = Files.newBufferedWriter(tmp)
+                    try writer.write(headersJson)
+                    finally writer.close()
+
+                    // Atomic move to the final location.
+                    Files.move(tmp, headersPath, StandardCopyOption.ATOMIC_MOVE)
+                    ()
+                  } else (),
+                  // If we couldn't get the lock, try again unless the file has been created.
+                  if (Files.exists(headersPath)) Some(())
+                  else None
+                )
+              }
+            }
+              .flatMap(_ => resolverMonad.fromTry(
+                Try(Files.readString(headersPath))
+                  .flatMap((new JawnParser).decode[Map[String, List[String]]](_) match {
+                    case Left(error) =>
+                      Failure(Recoverable(new RuntimeException(s"failed to parse headers file $headersPath", error)))
+                    case Right(obj) => Success(obj)
+                  })
+                  .map((headerMap) => HttpHeaders.of(headerMap.map { case (k, v) => (k, v.asJava) }.asJava, { (_, _) => true }))
+                  .flatMap { headers =>
+                    Try((
+                      headers
+                        .firstValue(digestType match {
+                          // See also https://maven.apache.org/resolver/expected-checksums.html#non-standard-x-headers
+                          case DigestType.Sha1 => "x-checksum-sha1"
+                          case DigestType.Sha256 => "x-checksum-sha256"
```

Comment:
I wonder if this should be moved to a method on `DigestType`? so this function is: `headers.firstValue(digestType.mavenHeader)` or something?

Also what about `x-goog-meta-checksum-sha1` and `x-goog-meta-checksum-md5`?

Maybe in fact we need to something like:

```scala
sealed trait DigestType {
  def getDigestInstance: MessageDigest
  def expectedHexLength: Int
  def name: String
  def headerKeys: List[String]
}
```
then do:

```scala
Try {
val digest = digestType
  .headerKeys
  .flatMap(key => headers.firstValue(key).asScala)
  .headOption
  .getOrElse(throw new Recoverable(new RuntimeException(s"no ${digestType} found in headers in $headersPath"))

val len = headers
  .firstValueAsLong("Content-Length")
  .orElseThrow(() =>
    Recoverable(new RuntimeException(s"no Content-Length found in headers $headersPath"))
  )

(digest, len)
}
```